### PR TITLE
Wait until tunnel is closed

### PR DESCRIPTION
**Pull request by MattiSG**

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -5,7 +9,7 @@ dependencies:
 
 test:
   override:
-    - cd sc-*-linux && ./bin/sc -u $SAUCE_USERNAME -k $SAUCE_ACCESS_KEY -f ~/sc_ready:
+    - cd sc-*-linux && ./bin/sc -u $SAUCE_USERNAME -k $SAUCE_ACCESS_KEY -f ~/sc_ready > $SAUCE_LOG:
```
Comment (d3ming) [inline circle.yml]: You can also use the -l option with sc:

```
cd sc-*-linux && ./bin/sc -u $SAUCE_USERNAME -k $SAUCE_ACCESS_KEY -f ~/sc_ready -l $SAUCE_LOG:
```

**Comment by bellkev** [comment]
Is there a specific benefit to waiting for the connection to close? It seems like this does introduce a bit of a risk of tying up circle containers for way too long if there's any issue with the log output.

**Comment by MattiSG** [comment]
> Is there a specific benefit to waiting for the tunnel to close?

Well, yes, having the tunnel actually close  ;)
If I don't add this code, my tunnels stay open. I think it is linked to the [teardown process](https://docs.saucelabs.com/reference/sauce-connect/#teardown-process):

> Once Sauce Connect is terminated (typically via ctrl-c), a call will be made from Sauce Connect to the REST API with instructions to terminate the tunnel VM. Sauce Connect will continue to poll the REST API until the tunnel VM has been halted and deleted.

Which I understand as “`sc` catches `SIGINT` and doesn't quit immediately”, and that the way Circle kills processes is more aggressive.

File: circle.yml (hunk 1 above)
Comment (MattiSG) [inline circle.yml]: True, but I thought a standard redirection to be more readable than yet another option you have to look up.

Comment (d3ming) [inline circle.yml]: :+1:  fair enough

**Comment by d3ming** [comment]
When I try out this change the while loop at the end waiting for 'Goodbye' just times out, it looks like SauceConnect just never tries to get killed.  What actually kills it at the end of the test?

Also as a more general question - is this how people actually run SauceConnect with CircleCI?  Start a new tunnel each time then tear it down at the end?  How will this scale with many different people running a CircleCI job with the same SauceLabs account?  AFAIK there is a 10 tunnel limit to the accounts

**Comment by MattiSG** [comment]
> What actually kills it at the end of the test?

Oh damn, you're right, I missed copying that part, sorry  :-/  Here's the [full file](https://github.com/sgmap/mes-aides-ui/blob/rewrite/circle.yml#L43-L46) I'm using (on two projects now), if you want to have a look. I'll add the missing `killall sc` here.

> is this how people actually run SauceConnect with CircleCI?

Good question. I don't know. But I do know that without this piece of code, my tunnels never closed, leading to failing builds as soon as the limit was hit.

> AFAIK there is a 10 tunnel limit to the accounts
It depends on your plan, but for the free plan it's actually 6, which makes it impossible to run two builds with 4x parallelism without closing all tunnels inbetween.

**Comment by d3ming** [comment]
@MattiSG current solution works =)  Only suggestion is to add a timeout to the while loop so it doesn't go on infinitely

File: circle.yml [hunk 2]
```diff
@@ -15,3 +19,7 @@ test:
     - curl --retry 10 --retry-delay 2 -v http://localhost:5000
     # Run selenium tests
     - nosetests
+
+  post:
+    - killall sc
+    - while ! grep -q 'Goodbye' $SAUCE_LOG; do sleep 1; done  # wait until Sauce Connect closes the tunnel
```
Comment (d3ming) [inline circle.yml]: Maybe add a ```timeout: 10``` so it doesn't go on forever (10min) in the event the log message changes or something else

Comment (MattiSG) [inline circle.yml]: Since it's quiet, Circle's standard timeout should be in effect.

**Comment by MattiSG** [comment]
Wow, just found a much simpler to do that with the `--wait` option of `killall`! I'll add a commit here, but it might be worth squashing them together.

File: circle.yml [hunk 3]
```diff
@@ -1,3 +1,7 @@
+machine:
+  environment:
+    SAUCE_LOG: $CIRCLE_ARTIFACTS/sauce_log.txt
```
Comment (MattiSG) [inline circle.yml]: This part not mandatory anymore with the latest `--wait` implementation. It is still nice to collect Sauce Connect's logs, but an intermediate variable might be overkill if used in only one place.

**Comment by MattiSG** [comment]
Ok, I'll open separate PRs, there are two separate concerns here. Closing.

**Pull request closed without merging**